earth_enterprise/src/common/FileAccessorPluginLoader_unittest.cpp

```diff
@@ -0,0 +1,180 @@
+  virtual int Close() {return 0;}
+  virtual bool PreadAll(void* buffer, size_t size, off64_t offset) {return false;}
+  virtual bool PwriteAll(const void* buffer, size_t size, off64_t offset) {return false;}
+  virtual bool ReadStringFromFile(const std::string &filename, std::string &str, std::uint64_t limit = 0) {return false;}
+  virtual bool Exists(const std::string &filename) {return false;}
+  virtual bool GetLinesFromFile(std::vector<std::string> &lines, const std::string &filename) {return false;}
+  virtual void fprintf(const char *format, ...) {}
+};
+
+class TestAccessor9876: public TestAccessorBase {
+public:
+  virtual int getFD() {return 9876;}
+};
+
+class TestAccessor5432: public TestAccessorBase {
+public:
+  virtual int getFD() {return 5432;}
+};
+
+
+//
+// FileAccessorFactory implementations and instances
+//
+class TestAccessorFactory9876: public FileAccessorFactory {
+protected:
+  virtual AbstractFileAccessor* GetAccessor(const std::string &fileName) override {
+   if (fileName == "9876file")
+    return new TestAccessor9876;
+   else return nullptr;
+  }
+};
+
+class TestAccessorFactory5432: public FileAccessorFactory {
+protected:
+  virtual AbstractFileAccessor* GetAccessor(const std::string &fileName) override {
+   if (fileName == "5432file")
+    return new TestAccessor5432;
+   else return nullptr;
+  }
+};
+
+static TestAccessorFactory9876 factory9876;
+static TestAccessorFactory5432 factory5432;
+
+
+//
+// Load and unload functions
+//
+void LoadPluginsTestFunc (const std::string &pluginDirectory, std::vector<FileAccessorPluginLoader::PluginFactory> &factories) {
+  factories.push_back({(void*)1, &factory9876});
+  factories.push_back({(void*)2, &factory5432});
+}
+
+void Load_NO_PluginsTestFunc (const std::string &pluginDirectory, std::vector<FileAccessorPluginLoader::PluginFactory> &factories) {
+}
+
+void UnloadPluginsTestFunc (FileAccessorPluginLoader::PluginHandle handle) {
+  // Do nothing. The default function calls dlclose on the handle.
+}
+
+
+//
+// Test plugin loaders
+//
+class TestPluginLoader: public FileAccessorPluginLoader {
+public:
+  TestPluginLoader(FileAccessorPluginLoader::LoadPluginFunc LoadImpl): FileAccessorPluginLoader(LoadImpl, UnloadPluginsTestFunc) {}
+};
+
+class TestPluginLoader_InvalidPluginDir: public FileAccessorPluginLoader {
+public:
+  // Passing in nullptr for the load/unload functions ensures that it will attempt
+  // to scan the directory for plugins.
+  TestPluginLoader_InvalidPluginDir(): FileAccessorPluginLoader(nullptr, nullptr, "/zzzz/thisisnot/areal/path/zzzz") {}
+};
+
+class TestPluginLoader_NoValidPlugins: public FileAccessorPluginLoader {
+public:
+  // Passing in nullptr for the load/unload functions ensures that it will 
+  // scan the directory for plugins. Passing in the current directory which
+  // should not contain any valid file accessor plugins.
+  TestPluginLoader_NoValidPlugins(): FileAccessorPluginLoader(nullptr, nullptr, "./") {}
+};
+
+
+//
```

Comment:
It might be nice to include a test that actually loads a fake plugin from disk.  Of course, that would be difficult to set up, but if we could get it working without too much effort it would improve our testing.

Reply:
I added a couple of new unit tests: one to test loading a valid plugin and one to test an attempt to load an executable file that isn't a shared object.
I also refactored the test that was attempting to load an invalid plugin to use a similar pattern.